sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -862,11 +933,45 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
     from match {
       case BinaryType =>
         (c, evPrim, evNull) => code"$evPrim = UTF8String.fromBytes($c);"
+      case DateType if format.isDefined =>
+        (c, evPrim, evNull) =>
+          if (customDateFormatter == null) {
+            code"""$evNull = true;"""
+          } else {
+            val formatterClass = classOf[DateFormatter]
+            val fmtr = JavaCode.global(
+              ctx.addReferenceObj("formatter", customDateFormatter, formatterClass.getName),
+              formatterClass)
+            code"""
+               |try {
+               |  $evPrim = UTF8String.fromString($fmtr.format($c));
+               |} catch (java.lang.IllegalArgumentException e) {
```

Comment:
Here can be a few more cases `DateTimeException` and `ArithmeticException`

Reply:
I don't think that `DateFormatter.format()` can throw `ArithmeticException` but `TimestampFormatter` can, so added there.  `DateTimeException` catch is added to both places.